src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -237,8 +234,78 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
                 queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type, _exportJobRecord.ResourceType));
             }
 
+            var filteredResources = new HashSet<string>();
+            if (_exportJobRecord.Filters != null)
+            {
+                foreach (var filter in _exportJobRecord.Filters)
+                {
+                    filteredResources.Add(filter.ResourceType);
+                }
+            }
+
             IAnonymizer anonymizer = IsAnonymizedExportJob() ? await CreateAnonymizerAsync(cancellationToken) : null;
 
+            if (!progress.RunFilteredSearches)
+            {
+                var index = -1;
+                if (_exportJobRecord.Filters != null)
+                {
+                    foreach (var filter in _exportJobRecord.Filters)
+                    {
+                        index++;
+
+                        if (progress.CurrentFilter != null && filter != progress.CurrentFilter)
+                        {
+                            continue;
+                        }
+
+                        if (_exportJobRecord.ExportType == ExportJobType.All || filter.ResourceType.Equals(KnownResourceTypes.Patient, StringComparison.Ordinal))
+                        {
+                            progress.SetFilter(filter);
+
+                            List<Tuple<string, string>> filterQueryParametersList = new List<Tuple<string, string>>(queryParametersList);
+                            foreach (var param in filter.Parameters)
+                            {
+                                filterQueryParametersList.Add(param);
+                            }
+
+                            await SearchWithFilter(exportJobConfiguration, progress, filter.ResourceType, filterQueryParametersList, sharedQueryParametersList, anonymizer, "filter" + index + "-", cancellationToken);
+
+                            progress.SetFilter(null);
+                        }
+                    }
+                }
+
+                progress.MarkFiltersFinished();
+                await UpdateJobRecordAsync(cancellationToken);
+            }
+
+            // This runs if the filteredResources doesn't contain Patient so that when a Patient or Group export is run it will run a full search of Patients.
+            // If Patients was filtered the search would already have been run above.
+            if (_exportJobRecord.ExportType == ExportJobType.All || !filteredResources.Contains(KnownResourceTypes.Patient))
+            {
+                foreach (var resource in filteredResources)
+                {
+                    queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type + ":not", resource));
```

Comment:
I am wondering whether the following might be easier to do compared to using `:not`. After running all the searches for the filtered resources, we can then run a single search for the rest of the resource types present in `_type`. That would mean we will have to parse the `_type` parameter values, compare with `filteredResources` (which we already have) and remove the common ones. We would get two advantages though:
1. No potential empty searches (if _type=MedicationRequest&_typeFilter=MedicationRequest%3Fstatus%3Dactive)
2. Should work for SQL as well since we don't have to use the `:not` modifier

Reply:
The issue I see is that the user isn't required to have the _type parameter if they have _typeFilter. I wish we didn't need `:not` since this story can't finish until `:not` is implemented.